### PR TITLE
Issue #540 - Fix building RPMs on ARM platforms.

### LINKED ISSUE #540: Unable to generate RPMs on ARM platforms.
rpmbuilder barfs when trying to build rpms on ARM. I have been able to manually workaround this by removing the 'BuildArch' line from the generated rpm.spec file, and adding '--target armv7l' to the rpmbuild call. I am not sure we technically need a BuildArch line, as we are not actually compiling as part of the rpm build.

```
root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/util.rb:101:in rescue in shellout!' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/util.rb:97:inshellout!'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/packagers/rpm.rb:354:in create_rpm_file' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/packagers/rpm.rb:76:inblock in <class:RPM>'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/packagers/base.rb:144:in instance_eval' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/packagers/base.rb:144:inrun!'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/project.rb:1062:in package_me' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/project.rb:1012:inbuild'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/cli.rb:83:in build' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/gems/thor-0.19.1/lib/thor/command.rb:27:inrun'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in invoke_command' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/gems/thor-0.19.1/lib/thor.rb:359:indispatch'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/cli/base.rb:33:in dispatch' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/gems/thor-0.19.1/lib/thor/base.rb:440:instart'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/lib/omnibus/cli.rb:41:in execute!' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bundler/gems/omnibus-41cfeee13cdb/bin/omnibus:11:in<top (required)>'
/root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bin/omnibus:23:in load' /root/omnibus-chef/.bundle/vendor/ruby/2.2.0/bin/omnibus:23:in<main>'
```

Removing the BuildArch line from the generated file, "/tmp/angrychef20150903-30121-1gtfne1/SPECS/angrychef-12.5.0~current.0+20150903040315.git.281.8143bf6-1.fedora22.armv7l.rpm.spec" and adding "--target armv7l" fixed the issue.

Fixed rpm command output:
```
# fakeroot rpmbuild --target armv7l -bb --buildroot /tmp/angrychef20150903-30121-1gtfne1/BUILD --define '_topdir /tmp/angrychef20150903-30121-1gtfne1' /tmp/angrychef20150903-30121-1gtfne1/SPECS/angrychef-12.5.0~current.0+20150903040315.git.281.8143bf6-1.fedora22.armv7l.rpm.spec 
Building target platforms: armv7l
Building for target armv7l
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.KnzKOz
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.YtGSdY
Executing(%install): /bin/sh -e /var/tmp/rpm-tmp.Ov5jjp
Processing files: angrychef-12.5.0~current.0+20150903040315.git.281.8143bf6-1.fedora22.armv7l
Checking for unpackaged file(s): /usr/lib/rpm/check-files /tmp/angrychef20150903-30121-1gtfne1/BUILD
Wrote: /tmp/angrychef20150903-30121-1gtfne1/RPMS/armv7l/angrychef-12.5.0~current.0+20150903040315.git.281.8143bf6-1.fedora22.armv7l.rpm
```

### PR DESCRIPTION
**Pull request by rhass**
The BuildArch line technically is not needed as we are not building anything during the RPM generation stage. On some systems this line causes a failure, which can be handled by specifying the target arch
when creating the rpm instead.

Closes #540.

### CODE REVIEW

File: spec/unit/packagers/rpm_spec.rb [hunk 1]
```diff
@@ -252,7 +251,7 @@ module Omnibus
 
       it 'uses the correct command' do
         expect(subject).to receive(:shellout!)
-          .with(/rpmbuild -bb --buildroot/)
+          .with(/rpmbuild --target x86_64 -bb --buildroot/)
         subject.create_rpm_file
       end
```
Comment (rhass) [inline spec/unit/packagers/rpm_spec.rb]: I think I probably should be stubbing out safe_architecture here to ensure that the target is always x86_64 in this test case.

**Comment by rhass** [comment]
I am not sure what to make of the travis results as it seems to have skipped the test we care about here:
https://travis-ci.org/chef/omnibus/jobs/78684967#L196

**Comment by rhass** [comment]
Manually validated (x86-64) build to ensure no regressions on CentOS with the following kitchen.local.yml and command:

```yaml
driver:
  name: docker
  username: vagrant
  password: vagrant

transport:
  username: vagrant

platforms:
  - name: centos
    driver_config:
      image: centos:6
```

```
bundle exec kitchen exec chef-centos -c 'source ~/load-omnibus-toolchain.sh && sudo yum -y install tar && rm -rf omnibus-chef ; git clone https://github.com/rhass/omnibus-chef.git && cd omnibus-chef && git checkout master && bundle install --without development --path .bundle/vendor && bundle exec omnibus build chef'
```

Result:

```
                    [Project: chef] I | Building version manifest
                      [HealthCheck] I | Running health on chef
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907072345.git.302.dc1686f' to `12.5.0~current.0+20150907072345.git.302.dc1686f'.
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907072345.git.302.dc1686f' to `12.5.0~current.0+20150907072345.git.302.dc1686f'.
                    [Packager::RPM] I | Rendering `/home/vagrant/omnibus-chef/.bundle/vendor/ruby/2.1.0/bundler/gems/omnibus-938a719074e5/resources/rpm/spec.erb' to `/tmp/chef20150907-37840-1u66jwi/SPECS/chef-12.5.0~current.0+20150907072345.git.302.dc1686f-1.el6.x86_64.rpm.spec'
                    [Packager::RPM] I | Creating .rpm file
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907072345.git.302.dc1686f' to `12.5.0~current.0+20150907072345.git.302.dc1686f'.
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907072345.git.302.dc1686f' to `12.5.0~current.0+20150907072345.git.302.dc1686f'.
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907072345.git.302.dc1686f' to `12.5.0~current.0+20150907072345.git.302.dc1686f'.
                       [Compressor] I | No compressor defined for `rhel'.
                    [Project: chef] I | Building version manifest
zlib(finalizer): the stream was freed prematurely.
rhass@rouge-ubuntu:~/src/omnibus-chef$  echo $?
0
```

As a side note, I used docker rather than vagrant due to resource constraints. 

Also manually validated the ARM build on Fedora 22 (CentOS not available on Scaleway. Manual build required as this work is bleeding edge and there is no other way to do this yet until my work is done here):

```
                    [Project: chef] I | Building version manifest
                      [HealthCheck] I | Running health on chef
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907225631.git.302.dc1686f' to `12.5.0~current.0+20150907225631.git.302.dc1686f'.
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907225631.git.302.dc1686f' to `12.5.0~current.0+20150907225631.git.302.dc1686f'.
                    [Packager::RPM] I | Rendering `/root/omnibus-chef/.bundle/vendor/ruby/2.1.0/bundler/gems/omnibus-938a719074e5/resources/rpm/spec.erb' to `/tmp/chef20150907-1844-1xearps/SPECS/chef-12.5.0~current.0+20150907225631.git.302.dc1686f-1.fedora22.armv7l.rpm.spec'
                    [Packager::RPM] I | Creating .rpm file
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907225631.git.302.dc1686f' to `12.5.0~current.0+20150907225631.git.302.dc1686f'.
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907225631.git.302.dc1686f' to `12.5.0~current.0+20150907225631.git.302.dc1686f'.
                    [Packager::RPM] W | Tildes hold special significance in the RPM package versions. They mark a version as lower priority in RPM's version compare logic. We'll replace all dashes (-) with tildes (~) so pre-releaseversions get sorted earlier then final versions. Converting`12.5.0-current.0+20150907225631.git.302.dc1686f' to `12.5.0~current.0+20150907225631.git.302.dc1686f'.
                       [Compressor] I | No compressor defined for `fedora'.
                    [Project: chef] I | Building version manifest
```

pkg/chef-12.5.0~current.0+20150907225631.git.302.dc1686f-1.fedora22.armv7l.rpm.metadata.json:
```json 
{
  "basename": "chef-12.5.0~current.0+20150907225631.git.302.dc1686f-1.fedora22.armv7l.rpm",
  "md5": "9a7865c325660721953bb2e58d521fee",
  "sha1": "ea1ad0e46db2c719d6ea7a66ec755c45607a2273",
  "sha256": "64f19c2a48d8a271331b4f1a98b3a0d5b5625a608b41e4fbaf997b9d4ae21919",
  "sha512": "e52d7ee187b256032461018748c185553eab4ab6ef233744fd9c3deeb564380a7bca5133cf5d14bc4d1d112ab0a6543d87ed0415f1aa23d5ee80e7f4ce3eff49",
  "platform": "fedora",
  "platform_version": "22",
  "arch": "armv7l",
  "name": "chef",
  "friendly_name": "Chef Client",
  "homepage": "https://www.chef.io",
  "version": "12.5.0-current.0+20150907225631.git.302.dc1686f",
  "iteration": 1
}
```

File: spec/unit/packagers/rpm_spec.rb (hunk 1 above)
Comment (schisamo) [inline spec/unit/packagers/rpm_spec.rb]: Something like this should work for you:

```diff
diff --git a/spec/unit/packagers/rpm_spec.rb b/spec/unit/packagers/rpm_spec.rb
index a73de3a..8c2d2bd 100644
--- a/spec/unit/packagers/rpm_spec.rb
+++ b/spec/unit/packagers/rpm_spec.rb
@@ -19,6 +19,7 @@ module Omnibus
     let(:project_root) { File.join(tmp_path, 'project/root') }
     let(:package_dir)  { File.join(tmp_path, 'package/dir') }
     let(:staging_dir)  { File.join(tmp_path, 'staging/dir') }
+    let(:architecture) { 'x86_64' }
 
     before do
       Config.project_root(project_root)
@@ -32,7 +33,9 @@ module Omnibus
       create_directory("#{staging_dir}/SOURCES")
       create_directory("#{staging_dir}/SPECS")
 
-      stub_ohai(platform: 'redhat', version: '6.5')
+      stub_ohai(platform: 'redhat', version: '6.5') do |data|
+        data['kernel']['machine'] = architecture
+      end
     end
 
     describe '#signing_passphrase' do
@@ -251,7 +254,7 @@ module Omnibus
 
       it 'uses the correct command' do
         expect(subject).to receive(:shellout!)
-          .with(/rpmbuild --target x86_64 -bb --buildroot/)
+          .with(/rpmbuild --target #{architecture} -bb --buildroot/)
         subject.create_rpm_file
       end
```

**Comment by danielsdeleo** [comment]
I'm not an expert on RPMs, the explanation for the changes makes sense to me, but I won't be able to tell you if there could be other side effects. The actual code change is good. I'd like to see @schisamo's suggestion for changes to the test implemented. :+1: aside from that.

File: spec/unit/packagers/rpm_spec.rb (hunk 1 above)
Comment (rhass) [inline spec/unit/packagers/rpm_spec.rb]: I tried it in a slightly different fashion, though I could have done it
wrong as I am still learning rspec:

https://github.com/chef/omnibus/commit/610869efd8b54f4f6b1877bb12a9bb901737ec6b

**Comment by sersut** [comment]
@rhass thanks for sending this PR. Researching this a bit with @yzl we believe this is the right direction to go. We are :+1: with the spec changes @schisamo identified. 

Your code is working fine. But @schisamo's example is more consistent with the rest of the spec code and also it also makes sure that the architecture information is coming from ohai. Let me know if that example is not clear and I can help you out.

**Comment by rhass** [comment]
No problem, I will update the rspec tonight with the suggested changes so
we can land this fix.

**Comment by sersut** [comment]
Awesome @rhass. Thanks.

**Comment by rhass** [comment]
Sorry for the delay -- I have been predisposed. I will try to do the rspec changes very soon.

I also did some additional testing and found there may need to be some other work done with respect to ARM for the safe_architecture bits. Fedora for example uses armv7hl -- where opensuse uses armv7l. :sob: 

I still think this patch itself is valid an useful, and in the interest of incremental value, I will likely make another patch to handle the various ways distros decide to see what ARM looks like. 

In the meantime -- if you want to see really something cool, try this out:

```yaml
driver:
  name: scaleway
  scaleway_org: <YOUR_ORG_TOKEN>
  scaleway_access_token: <YOUR_ACCESS_TOKEN>
provisioner:
  require_chef_omnibus: 12.5.0-current
  chef_omnibus_url: http://chef-arm.herokuapp.com/install.sh
platforms:
  - name: opensuse-13.2
```

Then of course do `kitchen converge chefdk-opensuse-132`

The heroku dyno running omnitruck might be sleeping if/when you try it, so give it a moment (or ping the install.sh before you run kitchen) -- it will work. I also haven't enabled cloudfront on the s3 bucket, so its a little slow to download Chef for the moment. Nevertheless, it's super awesome progress! :smile:

**Comment by sersut** [comment]
Thanks @rhass. Agreed. :+1: 

@chef/omnibus-maintainers can y'all check this and see if you can catch anything?

**Comment by danielsdeleo** [comment]
Looks ready to merge to me :+1:

**Comment by scotthain** [comment]
:+1:

**Comment by yzl** [comment]
:+1:

**Comment by sersut** [comment]
Thanks @rhass.

**Pull request merged**